tests/python/test_multidevice.py

```diff
@@ -112,3 +113,362 @@ def multidevice_schedule(self):
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear0_weight = self.define_tensor(
+            shape=[d, h * 4 // d, h],
+            contiguity=[True if d > 1 else None, True, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear0_bias = self.define_tensor(
+            shape=[d, h * 4 // d],
+            contiguity=[True if d > 1 else None, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear1_weight = self.define_tensor(
+            shape=[d, h, h * 4 // d],
+            contiguity=[True if d > 1 else None, True, True],
+            dtype=DataType.BFloat16,
+        )
+        self.mlp_linear1_bias = self.define_tensor(
+            shape=[h],
+            contiguity=[True],
+            dtype=DataType.BFloat16,
+        )
+        S17 = self.define_scalar(0.00000, dtype=DataType.Double)
+        S18 = self.define_scalar(1.00000, dtype=DataType.Double)
+        T23 = self.ops.uniform(
+            S17,
+            S18,
+            shape=[b, s, h],
+            rng_seed=S1,
+            rng_offset=S0,
+            dtype=DataType.BFloat16,
+        )
+        S24 = self.define_scalar(0.00000, dtype=DataType.Double)
+        S25 = self.define_scalar(1.00000, dtype=DataType.Double)
+        T30 = self.ops.uniform(
+            S24,
+            S25,
+            shape=[b, s, h],
+            rng_seed=S3,
+            rng_offset=S2,
+            dtype=DataType.BFloat16,
+        )
+        T31 = self.ops.cast(self.input, dtype=DataType.Float)
+        S32 = self.define_scalar(0.900000, dtype=DataType.Double)
+        T33 = self.ops.lt(T23, S32)
+        S34 = self.define_scalar(0.900000, dtype=DataType.Double)
+        T35 = self.ops.lt(T30, S34)
+        T36, T37 = self.ops.var_mean(T31, dims=[2], correction=0, keepdim=False)
+        T42 = self.ops.broadcast_in_dim(T36, shape=[b, s, 1], broadcast_dims=[0, 1])
+        T47 = self.ops.broadcast_in_dim(T37, shape=[b, s, 1], broadcast_dims=[0, 1])
+        S48 = self.define_scalar(1.00000e-05, dtype=DataType.Double)
+        T49 = self.ops.add(T42, S48)
+        T54 = self.ops.broadcast_in_dim(T47, shape=[b, s, h], broadcast_dims=[0, 1, 2])
+        T55 = self.ops.rsqrt(T49)
+        T56 = self.ops.sub(T31, T54)
+        T61 = self.ops.broadcast_in_dim(T55, shape=[b, s, h], broadcast_dims=[0, 1, 2])
+        T62 = self.ops.mul(T56, T61)
+        T67 = self.ops.broadcast_in_dim(
+            self.layernorm0_weight, shape=[b, s, h], broadcast_dims=[2]
+        )
+        T68 = self.ops.cast(T67, dtype=DataType.Float)
+        T69 = self.ops.mul(T62, T68)
+        T74 = self.ops.broadcast_in_dim(
+            self.layernorm0_bias, shape=[b, s, h], broadcast_dims=[2]
+        )
+        T75 = self.ops.cast(T74, dtype=DataType.Float)
+        T76 = self.ops.add(T69, T75)
+        T77 = self.ops.cast(T76, dtype=DataType.BFloat16)
```

Comment:
I understand this is a decomposed layer norm. Could LayerNorm be grouped in a separate function for readability?

Reply:
ditto